rfcs/0137-nix-language-version.md

```diff
@@ -0,0 +1,678 @@
+
+# Motivation
+[motivation]: #motivation
+
+The stability of Nix language has been praised on multiple occasions, e.g. [Nix and legacy enterprise software development: an unlikely match made in heaven](https://talks.nixcon.org/nixcon-2022/talk/QQPBFW/).
+Yet, as with any software system, in order to accommodate new insights, we want to allow the Nix language to evolve.
+This sometimes involves backward-incompatible ("breaking") changes that currently cannot be made without significant downstream disruption.
+
+Therefore we propose a mechanims and policies to introduce changes to the Nix language in a controlled and deliberate manner.
+It aims to avoid breaking existing code, to prevent inadvertently breaking reproducibility, and to minimise maintenance burden for implementors and users.
+
+## Motivating examples
+
+Incompatible changes from the past:
+
+- [A changelog of Nix (language) versions](https://code.tvl.fyi/about/tvix/docs/lang-version.md), as reflected in `builtins.langVersion`
+- There have been other, sometimes breaking changes to the language that have not resulted in an increment of the language version (e.g. the recent `fetchGit` changes).
+- The `builtins.toJSON 1.000001` output [changed in Nix 2.12](https://github.com/NixOS/nix/issues/8259).
+
+Possible future changes that are in discussion:
+
+- Remove URL literals (currently implemented via experimental-features)
+- Remove the old `let { body = ... }` syntax
+- Disallow leading zeroes in integer literals (such as `umask = 0022`)
+- Disallow `a.x or y` if `a` is not an attribute set
+- Simplifying semantics of `builtins.toString` and string interpolation
+- Remove the `let { body }` syntax
+- `__functor` and `__toString`, probably
+- Remove `__overrides`
+- [Make `builtins` more consistent](https://github.com/NixOS/nix/issues/7290), e.g. not exposing `map`, `removeAttrs`, `fetchGit` and and others in the global scope
+- Fix [imprecision in string representation of floating point numbers](https://github.com/NixOS/nix/pull/6238)
+- Make the `@`-pattern consistent with destructuring
+- A syntax to index into lists, e.g. `[ 1 2 3 ].0 == 1`
+- Use `,` to delimit elements of a list expression, like `[ 1, 2, 3 ]`
+- Do something about `?` meaning two different things depending on where it occurs (`{ x ? "" }:` vs `x ? ""`)
+- Better support for static analysis
+- [Syntax for hexadecimal numbers](https://github.com/NixOS/nix/pull/7695)
+
+Other discussions around language changes:
+
+- [Make path semantics less surprsing](https://github.com/NixOS/nix/issues/7338)
+- [RFC 110](https://github.com/NixOS/rfcs/pull/110) proposing a alternatives to the `with` expression
+- [Nix 2 – a hypothetical syntax](https://md.darmstadt.ccc.de/s/nix2)
+- [Nix language changes](https://gist.github.com/edolstra/29ce9d8ea399b703a7023073b0dbc00d)
+
+## Drawbacks
+
+Allowing multiple language versions to coexist will over time introduce a proliferation of syntax highlighters and other tooling.
```

Comment:
It also significantly complicates any evaluator implementation and makes comprehensive test coverage (which we do not have at the moment, I'd argue) harder.